Support freethreading Python

scoder opened the pull request:
See ticket https://bugs.launchpad.net/lxml/+bug/2111289

scoder committed:
Initial test with freethreaded Python.

scoder committed:
Release the GIL in FT-Python.

scoder committed:
First attempt at adding thread locks for tree manipulation.

scoder committed:
Initial refactoring to make the XML names dict local to a parser rather than local to a thread.

scoder committed:
Merge branch 'parser_dict' into ft_python

scoder committed:
Fix typo.

scoder committed:
Use bit field for parser config flags instead of a series of several oversized ints.

scoder committed:
Move lock from document to parser to include the (write) access to the parser's names dict.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Locking cannot fail, so mark all methods as 'void' and 'noexcept' .

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Minor code cleanups.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Tests and better lock implementation.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Rewrite locking based on atomics.

scoder committed:
Find better thread names for test debugging.

scoder committed:
Simplify read-lock handling for thread that own the write-lock.

scoder committed:
Fix some usage issues with "self._doc" in "_ElementTree" methods.

scoder committed:
Prevent document lock from disappearing during GC.

scoder committed:
Make sure we always have a lock for a parser.

scoder committed:
Remove outdated comment.

scoder committed:
Fix document reference for lock release.

scoder committed:
Use only a read-lock during Element deallocation by locking only the read-only proxy search. Once we know that there are no further proxies referring to the subtree. we don't need locking anymore.

scoder committed:
Postpone the locking in "Element.extend()" to fit it into the "lock_write_with()" order and avoid repeated locking of the same source document (since most appended elements will probably come from the same one).

scoder committed:
Improve debug output if assertion fails.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Add missing "unlock_read()" in "_Attrib.__bool__".

scoder committed:
Improve variable name.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Fix locking in _ElementMatchIterator.

scoder committed:
Use a "critical section" on the _Document in "_elementFactory()" instead of the per-parser lock, which is too broad.

scoder committed:
Fix typo.

scoder committed:
Remove an outdated work-around for a bug in Cython 0.28.

scoder committed:
Add locking wherever necessary.

scoder committed:
Lock documents of inserted elements in "_replaceSlice()".

scoder committed:
Add missing 'noexcept'  markers to public API.

scoder committed:
Add document locking functions to public C-API.

scoder committed:
Formatting.

scoder committed:
Update types in etreepublic.pxd to Py3.

scoder committed:
Add document locking to lxml.objectify.

scoder committed:
Make sure we deallocate the entire C subtree, including siblings, when deallocating an Element.

scoder committed:
Add a deferred cleanup mechanism if we fail to get the document lock on Element deallocation because other threads are writing concurrently. Instead, leave a reference for the current writer to clean up when its done.

scoder committed:
Avoid useless explicit C-API calls.

scoder committed:
Avoid coercion of the "atomic_int" type to Python int.

scoder committed:
Fix method name usage after renaming it.

scoder committed:
Formatting.

scoder committed:
Conservatively use a write lock instead of a read lock in XPath when users provide their own extensions.

scoder committed:
Fix uninitialised pointer usage.

scoder committed:
Conservatively use write lock instead of read lock on XSLT input documents if users provide extensions.

scoder committed:
Prevent C compiler optimisations in test to make sure threads get time to run in parallel.

scoder committed:
Simplify condition.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Improve readability in test.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Simplify conditional.

scoder committed:
Minor code cleanup.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Use atomics for replacing the proxy link pointer.

scoder committed:
Simplify proxy management code to only use a single, dedicated, short-lived lock.

scoder committed:
Tests: Enable tracemalloc in test runner.

scoder committed:
Improve test code readability.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Add a work-around for a Cython bug that prevents using PyMutex attributes in public cdef classes.

See https://github.com/cython/cython/issues/6995

scoder commented:
> if the check for reusing object code is `Py_REFCNT(op) == 1`, that check is wrong and needs updating to `_PyObject_IsUniquelyReferenced`

Thanks for the hint, but I'm not quite sure that's the right condition. `_PyObject_IsUniquelyReferenced()` (called by its public counterpart `PyUnstable_Object_IsUniquelyReferenced`) uses this code in the FT build:
```c
    return (_Py_IsOwnedByCurrentThread(ob) &&
            _Py_atomic_load_uint32_relaxed(&ob->ob_ref_local) == 1 &&
            _Py_atomic_load_ssize_relaxed(&ob->ob_ref_shared) == 0);
```
IIUC from [the PR that added it](https://github.com/python/cpython/pull/123272), `_Py_IsOwnedByCurrentThread(ob)` means that the object was created in the current thread and  is called here because the intention is to modify the object in the current thread. That's not the case for lxml. Instead, lxml wants to know if the object has been resurrected by a concurrent thread during its finalisation in the current thread, regardless of where it was created and who owns it. The check is guarded by a lock, and user code cannot hold a reference to the object any more (because it is being finalised), so there is not risk of concurrent refcount modification while we're evaluating the refcount check.

So, if `Py_REFCNT(op)` was not adapted to report the correct (and expected) refcount, how can I calculate the correct global refcount in lxml?

scoder commented:
> @scoder Btw if the check for reusing object code is Py_RECNT(op) == 1, that check is wrong and needs updating to _PyObject_IsUniquelyReferenced

Ah, I think what you meant was not the place where the proxy instance is deallocated but the place where the decision is taken whether to use an existing proxy instance as proxy for a C node.

In that case as well, the condition is wrong because it ensures a refcount of 1. A proxy instance can also be reused if the refcount is higher than that. (And I really don't care about how many threads create, use and throw away a proxy object – there can only ever be one, globally, for a given C node.)

The problem I've been fighting for a while now is that CPython can decide at any time to start the finalisation of a proxy object, and lxml will only learn about it when `__del__` is called. That's late because other threads might already have started to use the proxy while CPython is still preparing its deallocation. So `__del__` must be able to detect concurrent resurrection and stop the deallocation in that case.

I already added a refcount check to `hasProxy()`, which is used to determine if a proxy instance can be reused:
```cython
cdef inline bint hasProxy(xmlNode* c_node) noexcept:
    proxy = <cpython.object.PyObject*> c_node._private
    return (
        # Node has a proxy reference?
        proxy is not NULL and
        # Proxy reference is not currently deallocating?
        cpython.ref._Py_REFCNT(proxy) > 0
    )
```
https://github.com/scoder/lxml/blob/ft_python/src/lxml/proxy.pxi#L18-L27

But that still bares a (small but non-zero) risk of the refcount going to zero after checking the refcount with `hasProxy()` and before increfing it to use the proxy, so that CPython could still trigger the finalisation of an object that is used, which must then again be caught and aborted by `__del__`.

But, there's apparently a refcount check in `PyObject_CallFinalizerFromDealloc()`, which fails hard with a fatal C assertion if the refcount has been increased from 0 between the time it went down to 0 so that CPython started the finalisation and the time when something calls `PyObject_CallFinalizerFromDealloc()`. That is very annoying because it *can* happen and simply means that the object has been resurrected before it learned about its finalisation. And that seems perfectly ok to me, give that the finaliser is already allowed to resurrect to object.
I created a CPython ticket for this: https://github.com/python/cpython/issues/138101

Fidget-Spinner commented:
Sorry I pointed to the wrong function. It's `PyUnstable_Object_IsUniqueReferencedTemporary`, not `_PyObject_IsUniquelyReferenced`.

The difference is that if you have something on the stack, the refcount is no longer guaranteed to be the "true" refcount by CPython. CPython elides reference counts for certain temporaries on the stack now.

scoder commented:
The first thing that `PyUnstable_Object_IsUniqueReferencedTemporary` does is call `_PyObject_IsUniquelyReferenced`, which, as I argumented above, is already too restrictive.
https://github.com/python/cpython/blob/6fcac09401e336b25833dcef2610d498e73b27a1/Objects/object.c#L2745-L2766

Fidget-Spinner commented:
Yeah I get your point, I was answering this question:

> So, if Py_REFCNT(op) was not adapted to report the correct (and expected) refcount, how can I calculate the correct global refcount in lxml?

Basically, Py_REFCNT is no longer "correct" on 3.14 and later.

Fidget-Spinner commented:
Though I don't think `Py_REFCNT` is the problem here.

scoder committed:
Use TryIncRef in frethreading Python 3.14 to resurrect proxy objects.

scoder committed:
Make test a little less prone to long waits on deadlocks.

scoder committed:
Temporarily disable two tests in freethreading Python that do not work yet.

scoder committed:
Avoid using tracemalloc in PyPy, seems to be broken there.

scoder committed:
Try to fix a C warning  about a C warning.

scoder committed:
Build: Exclude FT-Python 3.13t, at least for now.

scoder committed:
Use atomic pointer type in helper function (and make sure it gets inlined).

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Remove unused atomic pointer declarations.

scoder committed:
CI: Remove 3.13t since it's probably not going to work due to missing C-API features (that were only added in Py3.14).

ngoldbaum commented:
@scoder if you'd like any help finishing this off, someone on our team at Quansight might be able to help. Just throwing that out there.

scoder committed:
Merge branch 'master' into ft_python

scoder commented on src/lxml/rwlock.pxi:
```diff
@@ -0,0 +1,318 @@
+    @cython.critical_section
+    cdef void unlock_write(self) noexcept:
+        assert self._write_locked_id == self._my_lock_id(), f"{self._write_locked_id} != {self._my_lock_id()}"
+        assert self._reader_count < 0, <nonatomic_int> self._reader_count
```
It's an assertion for something that must be globally true, regardless of concurrency. The thread sanitizer may be unhappy with it, but it's not an actual issue.

scoder commented on src/lxml/rwlock.pxi:
```diff
@@ -0,0 +1,318 @@
+        if self._writers_waiting == 0:
+            self._writers_wait_lock.release()
+
+    @cython.critical_section
```
> Using critical section here is not correct here because the underneath lock acquisition can release the thread state thereby releasing the critical section.

I had taken that into account and I think it's fine. The locking may do that, assuming that we re-enter the critical section after acquiring or releasing the lock. The critical section is mostly intended to guard the conditionals running up to the point where we determine that we need to lock. From that point on, things should become serialised by the write lock.

scoder committed:
Use only atomics and PyMutex, no critical sections.

scoder committed:
Fix incorrectly copied names.

scoder committed:
CI: Show C atomics variant in build logs.

scoder committed:
Build: Build wheels with C11 to use C atomics.

scoder committed:
CI: Download dependencies only once per platform, not once per job.

scoder committed:
Revert "Build: Build wheels with C11 to use C atomics."

This reverts commit 81754b7d5d5b6dc08cf4ab7ee2994a9ca03d9a38.

scoder committed:
Build: Use C11 for wheel and CI builds.

scoder commented:
Looks like we're close. There's a concurrent XSLT test that hangs which I'll have to investigate. I didn't pay much attention to XSLT so far and just schematically added locks there, so the failure isn't surprising.

Quite a bit of the rest seems to work. If anyone feels ambitious, you can give it a first try, without high expectations. You might want to use Cython 3.2 beta for the build, just in case. I didn't try it with 3.1 yet.

scoder committed:
Enable a working test.

scoder committed:
Fix an "uninitialised use on (unlikely) error" bug.

scoder committed:
Make reading "_reader_count" atomic and guard some accesses to "_write_locked_id".

scoder committed:
Modernise import in test.

scoder committed:
Fix build in Win32 that lacks "_InterlockedExchangeAdd64".

scoder committed:
Build: Add a makefile target for ThreadSanitizer builds.

scoder committed:
Rewrite "RWLock" to use atomics for the non-congested case, "PyMutex" for waiting, and "critical_section" for short bits of synchronisation.

scoder committed:
Comment out atomics functions that are currently unused.

scoder committed:
Test: Hide debug prints.

scoder committed:
Release the GIL for waiting.

scoder committed:
Simplify the lock implementation considerably by always locking the writer and using a simple counter protocol to unblock the readers.

scoder committed:
Improve test output on failure.

scoder committed:
Docs: Fix doctest code hiding.

scoder committed:
Docs: Add comment that recent libxml2 versions actually focus on HTML5.

scoder committed:
CI: Use regular Py3.14 builds and include 3.15-dev.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
CI: Enable benchmarks to compare original with locking lxml.

scoder committed:
Tests: Speed up the objectify fuzzer test a little.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Build: Use Cython 3.2.0.

scoder committed:
CI: Remove unused env-var.

scoder committed:
Build: Upgrade cibuildwheel to 3.2.1.

bebras123 commented:
Hello, thanks for your work. I tried a small script with lxml from https://github.com/lxml/lxml/actions/runs/19095197185. The script simply parses a large XML file (about 100MB) and runs a few XPath queries from multiple threads.

I noticed that the concurrent part of my script runs about twice as slow when using Python 3.14 or Python 3.14t together with lxml 7.0.0a, compared to Python 3.14 with lxml 6.0.2.

My full test setup is here: https://github.com/bebras123/lxml-python314t-test  
Test script: https://github.com/bebras123/lxml-python314t-test/blob/main/run.py

scoder committed:
Make XPath evaluation actually use read-locking instead of write-locking.
It accidentally used the write lock because it didn't notice that the registered regex extensions were safe, internal functions and not user provided ones (that might require write-locking).

scoder committed:
Provide lock performance counters to analyse the locking behaviour.

scoder committed:
Let the lock decide itself if it needs to release the GIL (for waiting).

scoder committed:
Add docstring to new "ElementTree.get_lock_perf_counters" method.

scoder committed:
Remove "try_lock_read" functionality.
It's brittle to allow retracting a read request while writers potentially respond concurrently to it.

scoder committed:
Add some comments.

scoder committed:
Fix unblocking of readers when a writer finishes.
Only the first reader must acquire the lock twice to block the other readers and to let the writer unblock it. All later readers can run through after the writer unblocked the first reader.

scoder committed:
Add a trace helper to log the thread interactions before deadlocks.

scoder committed:
Benchmarks: Include "libs" directory with downloaded library archives in cross-benchmark cache.

scoder committed:
Switch to Cython 3.2.2.

scoder committed:
Remove bug work-around that is no longer needed with Cython 3.2.

scoder committed:
Prefer relaxed C11 atomics over strict (add) PyAtomics since we don't care about ordering.

scoder committed:
Fix debug warning on MSVC.

scoder committed:
Upgrade cibuildwheel to 3.3.0.

scoder committed:
Improve thread name output in debugger.

scoder committed:
Fix C warning syntax.

scoder committed:
Reorder fast path.

scoder committed:
Fix argument names in declaration.

scoder committed:
Use an additional lock to safely start up all pending readers when a writer finishes.

scoder committed:
Reduce test runtime.

scoder committed:
Minor test cleanups.

scoder committed:
CI: Increase test timeout to avoid failures.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
CI: Remove merge duplication.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Set the 'CYTHON_UNSAFE_IGNORE_PYMUTEX_ABI_COMPATIBILITY' C macro to disable the safety check when automatically switching to "PyThread_type_lock" in abi3 wheels.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Minor code cleanup.

scoder committed:
Also consider globally registered XPath functions and XSLT extensions when deciding between read and write lock.

scoder committed:
Always compile the test lock module with LXML_LOCK_PERFORMANCE=1.

scoder committed:
Build: Unset "freethreading_compatible" on Py3.13 (currently crashes in Py3.13t).

scoder committed:
Build: Don't set "freethreading_compatible" directive at all in Py<3.14.

scoder committed:
CI: Add PyDebug test jobs for regular and freethreading Python main branch.

scoder committed:
CI: Trigger CI workflow on changes to PyDebug build workflow.

scoder committed:
CI: Pass environment config into "compiled_python" workflow.

scoder committed:
Build: Wrap config line into separate options.

scoder committed:
CI: Pass env vars as input into sub-workflow.

scoder committed:
CI: Pass env vars as input into sub-workflow.

scoder committed:
CI: Fix typo.

scoder committed:
CI: Pass "freethreading" flag as input into sub-workflow.

scoder committed:
CI: Set "STATIC_DEPS=true" also for PyDebug job.

scoder committed:
Fix some PyPy issues.

scoder committed:
CI: Fix library caching in PyDebug test setup.

scoder committed:
CI: Fix PyDebug dependency.

scoder committed:
CI: Avoid downloading CPython git history in PyDebug build.

scoder committed:
Tests: Try to get a more informative error message from a threading test.

scoder committed:
Rewrite the input resolver setup to make use of the new parser context local setting in libxml2 2.14.
This avoids global settings that are not thread-safe.

scoder committed:
Add an attribute to determine if freethreading support is properly enabled in the module.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Benchmarks: Report FT-Python versions with a 't' suffix.

scoder committed:
Add a benchmarks workflow to compare the feature branch to the master branch, and FT-Python to regular Python.

scoder committed:
Benchmarks: Fix build step condition.

scoder committed:
Fix typo.

scoder committed:
Benchmarks: Add missing CI file trigger.

scoder committed:
Benchmarks: Always copy the cached libraries from the project directory.

clin1234 reviewed (approved):
As I have no prior experience with testing `PyPy`, LGTM

scoder committed:
Turn the private C macro "__lxml_DEBUG_ATOMICS" into a public macro "LXML_DEBUG_ATOMICS" to make atomics easily debuggable for users.

scoder committed:
Benchmarks: Help with debugging build problems by making it easy to see the compiler output.

scoder committed:
Benchmarks: Report results in Github Actions as Markdown tables.

scoder committed:
Benchmarks: Keep baseline run first.

scoder committed:
CI: Fix some file triggers.

scoder committed:
Remove a bit of runtime overhead.

scoder committed:
Benchmarks: Reduce runtime of some etree benchmarks.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Benchmarks: Add missing file triggers.

scoder committed:
Benchmarks: Fix Limited API build.

scoder committed:
Benchmarks: Reduce repetitions to speed up the benchmark runs.

scoder committed:
Benchmarks: Exclude Limited API, currently crashes.

scoder committed:
Benchmarks: Fix time unit in markdown reports.

scoder committed:
Benchmarks: Upload CSV timings.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Tests: Make test base classes play more nicely with pytest.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

kumaraditya303 commented:
@scoder I tried running the `test_threading` tests under TSAN it is consistently hangs for me. There is also a data race in RWLock implementation which might be relevant. 

```console
==================
WARNING: ThreadSanitizer: data race (pid=50148)
  Write of size 8 at 0x000300c2e1e8 by thread T4:
    #0 __pyx_f_4lxml_5tests_9_testlock_6RWLock_unlock_write _testlock.c:4650 (_testlock.cpython-314t-darwin.so:arm64+0xd624)
    #1 __pyx_pw_4lxml_5tests_9_testlock_7_RWLock_17unlock_write _testlock.c:6412 (_testlock.cpython-314t-darwin.so:arm64+0x114ec)
    #2 __Pyx_CyFunction_Vectorcall_FASTCALL_KEYWORDS _testlock.c:13813 (_testlock.cpython-314t-darwin.so:arm64+0x15574)
    #3 PyObject_Vectorcall call.c:327 (libpython3.14t.dylib:arm64+0x8af00)
    #4 _PyEval_EvalFrameDefault generated_cases.c.h:1619 (libpython3.14t.dylib:arm64+0x27cc00)
    #5 _PyEval_Vector ceval.c:2083 (libpython3.14t.dylib:arm64+0x278928)
    #6 _PyFunction_Vectorcall call.c (libpython3.14t.dylib:arm64+0x8b54c)
    #7 method_vectorcall classobject.c:73 (libpython3.14t.dylib:arm64+0x8f9c0)
    #8 context_run context.c:728 (libpython3.14t.dylib:arm64+0x2c3d18)
    #9 method_vectorcall_FASTCALL_KEYWORDS descrobject.c:421 (libpython3.14t.dylib:arm64+0xa21cc)
    #10 PyObject_Vectorcall call.c:327 (libpython3.14t.dylib:arm64+0x8af00)
    #11 _PyEval_EvalFrameDefault generated_cases.c.h:1619 (libpython3.14t.dylib:arm64+0x27cc00)
    #12 _PyEval_Vector ceval.c:2083 (libpython3.14t.dylib:arm64+0x278928)
    #13 _PyFunction_Vectorcall call.c (libpython3.14t.dylib:arm64+0x8b54c)
    #14 method_vectorcall classobject.c:73 (libpython3.14t.dylib:arm64+0x8f9c0)
    #15 _PyObject_Call call.c:348 (libpython3.14t.dylib:arm64+0x8b1c0)
    #16 PyObject_Call call.c:373 (libpython3.14t.dylib:arm64+0x8b234)
    #17 thread_run _threadmodule.c:359 (libpython3.14t.dylib:arm64+0x419b3c)
    #18 pythread_wrapper thread_pthread.h:242 (libpython3.14t.dylib:arm64+0x369a20)

  Previous read of size 8 at 0x000300c2e1e8 by thread T3:
    #0 __pyx_f_4lxml_5tests_9_testlock_6RWLock_lock_write _testlock.c:4312 (_testlock.cpython-314t-darwin.so:arm64+0xd3a4)
    #1 __pyx_pw_4lxml_5tests_9_testlock_7_RWLock_15lock_write _testlock.c:6333 (_testlock.cpython-314t-darwin.so:arm64+0x113ac)
    #2 __Pyx_CyFunction_Vectorcall_FASTCALL_KEYWORDS _testlock.c:13813 (_testlock.cpython-314t-darwin.so:arm64+0x15574)
    #3 PyObject_Vectorcall call.c:327 (libpython3.14t.dylib:arm64+0x8af00)
    #4 _PyEval_EvalFrameDefault generated_cases.c.h:1619 (libpython3.14t.dylib:arm64+0x27cc00)
    #5 _PyEval_Vector ceval.c:2083 (libpython3.14t.dylib:arm64+0x278928)
    #6 _PyFunction_Vectorcall call.c (libpython3.14t.dylib:arm64+0x8b54c)
    #7 method_vectorcall classobject.c:73 (libpython3.14t.dylib:arm64+0x8f9c0)
    #8 context_run context.c:728 (libpython3.14t.dylib:arm64+0x2c3d18)
    #9 method_vectorcall_FASTCALL_KEYWORDS descrobject.c:421 (libpython3.14t.dylib:arm64+0xa21cc)
    #10 PyObject_Vectorcall call.c:327 (libpython3.14t.dylib:arm64+0x8af00)
    #11 _PyEval_EvalFrameDefault generated_cases.c.h:1619 (libpython3.14t.dylib:arm64+0x27cc00)
    #12 _PyEval_Vector ceval.c:2083 (libpython3.14t.dylib:arm64+0x278928)
    #13 _PyFunction_Vectorcall call.c (libpython3.14t.dylib:arm64+0x8b54c)
    #14 method_vectorcall classobject.c:73 (libpython3.14t.dylib:arm64+0x8f9c0)
    #15 _PyObject_Call call.c:348 (libpython3.14t.dylib:arm64+0x8b1c0)
    #16 PyObject_Call call.c:373 (libpython3.14t.dylib:arm64+0x8b234)
    #17 thread_run _threadmodule.c:359 (libpython3.14t.dylib:arm64+0x419b3c)
    #18 pythread_wrapper thread_pthread.h:242 (libpython3.14t.dylib:arm64+0x369a20)


SUMMARY: ThreadSanitizer: data race _testlock.c:4650 in __pyx_f_4lxml_5tests_9_testlock_6RWLock_unlock_write
==================

```

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Merge branch 'master' into ft_python

scoder commented:
For some reason, I'm getting clean CI runs in Py3.14t+ now, e.g.

https://github.com/lxml/lxml/actions/runs/24598027352/job/71932553703

https://github.com/lxml/lxml/actions/runs/24598027352/job/71932553818

It might be that the load of fixes elsewhere (Cython and CPython) made this work, but I'm happy to take it as a given.

The benchmarks show clear regressions, but mixed and somewhere between "not too bad for a start" and "needs investigation". We probably have to live with some of the slowdowns for the now generally improved safety, but there might be ways to keep the effects more limited.
https://github.com/lxml/lxml/actions/runs/24598027348?pr=477

PyPy also still needs fixing. I'll look into that soon.

Once again, if someone wants to try out the current state, please do and report back. Suitable wheels can be found here:
https://github.com/lxml/lxml/actions/runs/24598027357

scoder committed:
Add a lock to the Feed-Parser interface to prevent concurrent access.
It seems a bit useless given the intent to incrementally parse a sequential stream of data, but it's not impossible to find use cases where chunks come from different sources and ordering does not matter, which now allows concurrently feeding them in.

da-woods commented on src/lxml/rwlock.pxi:
```diff
@@ -0,0 +1,439 @@
+                        !defined(__STDC_NO_ATOMICS__) && \
+                       ATOMIC_INT_LOCK_FREE == 2)
+    /* C11 atomics are available and  ATOMIC_INT_LOCK_FREE is definitely on */
+    /* Prefer this over pyatomics.h, which enforces strict memory ordering. We only need atomic operations. */
```
I don't think this can be right. To implement a lock I think you really need at least `acquire` ordering on the operations that acquire the lock and `release` ordering on the operations that release the lock.

With `relaxed` ordering the compiler is allowed to move stuff done inside the lock body before the atomic operation that acquires it, or after the atomic operation that releases it. It only really guarantees that the number seen definitely happened at some point.

scoder replied:
You're probably right. I'll just follow what `pyatomics.h` does and require sequential consistency if the Python header is not available.

scoder committed:
Use strict sequential ordering for the rwlock.
Suggested by @da-woods.

scoder committed:
Prevent race conditions in objectify.set_default_parser().
Found by @devdanzin.

scoder committed:
Prevent race conditions in namespace registries.
Found by @devdanzin.

scoder committed:
Prevent race conditions when using and changing the global element class lookup scheme.
Found by @devdanzin.

scoder committed:
Prevent deadlock when re-entering parser.
Found by @devdanzin.

scoder committed:
Remove unnecessary C-ism.

scoder committed:
Use write-lock instead of read-lock for SubElement().
Found by @devdanzin.

scoder committed:
Use write-lock instead of read-lock for setting Entity names.
Found by @devdanzin.

scoder committed:
Use critical-section for setting Element line number.
Found by @devdanzin.

scoder committed:
Guard tag caching by write-lock to make sure the dict doesn't change before traversing the tree.
Found by @devdanzin.

scoder committed:
Guard XSLT serialisation with locks and its buffer handling with a critical section to make it thread-safe.
Found by @devdanzin.

scoder committed:
Avoid running XSLT extension registrations from empty dict.

scoder committed:
Simplify some code to make it clear that we always close the output buffer.

scoder committed:
Make parser context creation thread-safe.
Found by @devdanzin.

scoder committed:
Make lazy Python tag name initialisation thread-safe.
Found by @devdanzin.

scoder committed:
Move node evaluation into lock guards.

scoder committed:
Reduce inlining scope of cacheTags() method.

scoder committed:
Fix Cython fallback in diff.py.

scoder committed:
Disable finaliser usage in PyPy to work around crashes.

scoder committed:
Update issue link.

scoder committed:
Use existing dedicated flag instead of repeating condition.
Found by @devdanzin.

scoder committed:
Add unnecessary try-finally for consistency, in case we do start raising exceptions in the try-block.
Cython optimises away the error case if nothing can raise.

scoder committed:
Revert "Use existing dedicated flag instead of repeating condition."

This reverts commit 1544a6151c37bbd4b7a0ec081444c926804da7f4.

It seems that both are not equivalent in this case.

scoder committed:
Merge branch 'master' into ft_python

scoder committed:
Fix exception message.

scoder committed:
Mark and skip tests that currently crash in PyPy.

scoder committed:
Merge branch 'master' into ft_python

scoder commented:
I'll merge this PR once CI is happy and then start preparing an alpha release. PyPy is still in a sad state with a couple of hard test crashes, but those seem mostly limited to threading issues. I'll have another look at them later.

scoder commented:
Released as part of lxml 7.0.0a1.
